contributors/design-proposals/sig-cli/decoupling_kubectl_from_k8s.md

```diff
@@ -0,0 +1,244 @@
+
+This can wait.  For purposes of bootstrapping a new style of e2e testing and
+release, its convenient to already have a _decoupled_ kubectl in the wild,
+released per the _existing_ process.  Any new process will need a rollback state
+it understands.
+
+#### Moving kubectl to its [own repo][kubectl repo], outside the [k8s repo].
+
+This decoupling work is a precondition to such a move, but doesn't _require_ a
+new repo.
+
+## User Experience
+
+The user should see no change other than, it's hoped, faster feature development
+and bug fixes accruing from the kubectl/k8s decoupling.
+
+## Implementation
+
+The identification of bad dependencies, and the eventual prevention of their
+re-introduction will be accomplished via _visibility_ rules in `BUILD` files.
+Violations in visibility cause `make bazel-build` to fail, which in turn causes
+the submit queue to fail - so attempting to introduce a bad dependency is like a
+failing unit tests.
+
+### Background Facts
+
+ * A _package_ is any directory that contains a BUILD file.
+ 
+ * A `package_group` is a BUILD file term that defines a set of
+   packages for use in other rules.
+ 
+ * A _visibility rule_ takes a list of package groups as its
+   argument - or one of the pre-defined groups `//visibility:private` or
+   `//visibility:public`.
+ 
+ * If no visibility is explicitly defined, a package is _private_ by default.
+ 
+ * kubernetes has explicitly labeled its nearly 2000 packages as public - a
+   _sensible thing to do_ when first introducing bazel.
+   
+   Now that kubernetes builds under bazel, its possible to use bazel to
+   discover, fix and prevent bad deps.
+
+
+### Step One: Introduce visibility rules and clarify dependence on kubectl.
+
+When this work in this section is complete, all dependencies on kubectl code by
+non-kubectl code will either be removed or specifically allowed.
+
+ * A new k8s package will be created called `kubernetes/visible_to`.
+
+   It will have only two files: `BUILD` and `OWNERS`.
+
+ * The only rules in `BUILD` will define package groups.
+
+   These groups will exist only to be used in visibility rules in
+   other packages throughout k8s.
+
+ * All kubectl code will switch to private by default.
+
+   This means removal of all package-level policies like
+   ```
+   package(default_visibility = ["//visibility:public"])
+   ```
+
+ * Exceptions to this will be per-target.
+
+   A target will let others depend on it like this:
+   ```
+   visibility = ["//visible_to:kubectl_client"]
+   ```
+   Once all these changes are made, all dependencies are
+   intentional, albeit both good and bad.
+ 
+ * `OWNERS` file imposes special approval for visibility changes.
+
+   Review will require familiarity with visibility usage, since
+   small changes can open large holes.  This is intended to stop
+   introduction of new bad depenencies.
+
+ * Bad dependencies labelled for removal.
+
+   Bad deps will be allowed and identified with a special package
```

Comment:
I can imagine wanting to vendor kubectl into the main repo (forever) for full e2e testing.